### PR TITLE
Update the LLM Agents guide with Gemini2.0 Thinking demo

### PR DESCRIPTION
**Pull request by yvrjsharma**
## Description

Added Guide for Gemini2.0 Flash Thinking. 

Closes: #(issue)

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Testing and Formatting Your Code

1. PRs will only be merged if tests pass on CI. We recommend at least running the backend tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the backed tests: `bash scripts/run_backend_tests.sh`

2. Please run these bash scripts to automatically format your code: `bash scripts/format_backend.sh`, and (if you made any changes to non-Python files) `bash scripts/format_frontend.sh`

### CODE REVIEW

File: guides/05_chatbots/03_agents-and-tool-usage.md [hunk 1]
```diff
@@ -3,14 +3,14 @@
 Tags: LLM, AGENTS, CHAT
 Related spaces: https://huggingface.co/spaces/gradio/agent_chatbot, https://huggingface.co/spaces/gradio/langchain-agent
 
-The Gradio Chatbot can natively display intermediate thoughts and tool usage. This makes it perfect for creating UIs for LLM agents. This guide will show you how.
+The Gradio Chatbot can natively display intermediate thoughts and tool usage. This makes it perfect for creating UIs for LLM agents and Thinking LLMs. This guide will show you how.
```
Comment (abidlabs) [inline guides/05_chatbots/03_agents-and-tool-usage.md:6]:
```suggestion
The Gradio Chatbot can natively display intermediate thoughts and tool usage. This makes it perfect for creating UIs for LLM agents and chain-of-thought (CoT) demos. This guide will show you how to display intermediate thoughts.
```

File: guides/05_chatbots/03_agents-and-tool-usage.md [hunk 2]
```diff
@@ -166,4 +167,139 @@ demo.launch()
 That's it! See our finished langchain demo [here](https://huggingface.co/spaces/gradio/langchain-agent).
 
 
+## Building with Visible Thinking LLMs
```
Comment (abidlabs) [inline guides/05_chatbots/03_agents-and-tool-usage.md:170]:
```suggestion
## Chain-of-Thought (CoT) demos
```

File: guides/05_chatbots/03_agents-and-tool-usage.md [hunk 3]
```diff
@@ -166,4 +167,139 @@ demo.launch()
...
+
+
+The Gradio Chatbot can natively display intermediate thoughts of a _thinking_ LLM. This makes it perfect for creating UIs that show how an AI model "thinks" while generating responses. Below guide will show you how to build a chatbot that displays Gemini AI's thought process in real-time.
+
+
+### A real example using Gemini2.0 Flash Thinking API
```
Comment (abidlabs) [inline guides/05_chatbots/03_agents-and-tool-usage.md:176]:
```suggestion
### A real example using Gemini 2.0 Flash Thinking API
```

File: guides/05_chatbots/03_agents-and-tool-usage.md [hunk 4]
```diff
@@ -166,4 +167,139 @@ demo.launch()
...
+
+Let's create a complete chatbot that shows its thoughts and responses in real-time. We'll use Google's Gemini API for accessing Gemini2.0 Flash Thinking LLM and Gradio for the UI.
```
Comment (abidlabs) [inline guides/05_chatbots/03_agents-and-tool-usage.md:178]:
```suggestion
Let's create a complete chatbot that shows its thoughts and responses in real-time. We'll use Google's Gemini API for accessing Gemini 2.0 Flash Thinking LLM and Gradio for the UI.
```

File: guides/05_chatbots/03_agents-and-tool-usage.md [hunk 5]
```diff
@@ -166,4 +167,139 @@ demo.launch()
...
+
+We'll begin with imports and setting up the gemini client. Note that you'll need to acquire a Google Gemini API key first -
```
Comment (abidlabs) [inline guides/05_chatbots/03_agents-and-tool-usage.md:180]: Perhaps you can link to where one can obtain a Gemini API key
```suggestion
We'll begin with imports and setting up the gemini client. Note that you'll need to acquire a Google Gemini API key first -
```

File: guides/05_chatbots/03_agents-and-tool-usage.md [hunk 6]
```diff
@@ -166,4 +167,139 @@ demo.launch()
...
+            messages[-1] = ChatMessage(
+                role="assistant",
+                content=thought_buffer,
+                metadata={"title": "⏳Thinking: *The thoughts produced by the Gemini2.0 Flash model are experimental"}
+            )
+        
+        yield messages
+```
+
+Then, let's create the Gradio interface:
```
Comment (abidlabs) [inline guides/05_chatbots/03_agents-and-tool-usage.md:258]: Could we use `gr.ChatInterface` instead? `gr.ChatInterface` now supports the function returning multiple messages (you just return or yield a `list`), which would simplify the Gradio code and provide a better UI.

File: guides/05_chatbots/03_agents-and-tool-usage.md [hunk 7]
```diff
@@ -166,4 +167,139 @@ demo.launch()
...
+demo.launch()
+```
+
+This creates a chatbot that:
+
+- Displays the model's thoughts in a collapsible section
+- Streams the thoughts and final response in real-time
+- Maintains a clean chat history
+
+ That's it! You now have a chatbot that not only responds to users but also shows its thinking process, creating a more transparent and engaging interaction. See our finished Gemini2.0 Flash Thinking demo [here](https://huggingface.co/spaces/ysharma/Gemini2-Flash-Thinking).
```
Comment (abidlabs) [inline guides/05_chatbots/03_agents-and-tool-usage.md:304]:
```suggestion
 That's it! You now have a chatbot that not only responds to users but also shows its thinking process, creating a more transparent and engaging interaction. See our finished Gemini 2.0 Flash Thinking demo [here](https://huggingface.co/spaces/ysharma/Gemini2-Flash-Thinking).
```

**Review by abidlabs — APPROVED** [review]
LGTM @yvrjsharma left some suggestions but nothing major

File: guides/05_chatbots/03_agents-and-tool-usage.md (hunk 2 above)
Comment (yvrjsharma) [reply guides/05_chatbots/03_agents-and-tool-usage.md:170]: I believe the terms 'Thinking' and 'Reasoning' will be increasingly popular for LLMS. Therefore, I want to ensure that our messaging is aligned. In my opinion, 'thinking' is currently more commonly used than 'Chain-of-thought.' In fact, Google has referred to its model as 'Gemini2.0 Flash Thinking Experimental,' so adopting 'Thinking' seems more logical to me.

File: guides/05_chatbots/03_agents-and-tool-usage.md (hunk 5 above)
Comment (yvrjsharma) [reply guides/05_chatbots/03_agents-and-tool-usage.md:180]: Good idea! Will add a hyperlink.

File: guides/05_chatbots/03_agents-and-tool-usage.md (hunk 6 above)
Comment (yvrjsharma) [reply guides/05_chatbots/03_agents-and-tool-usage.md:258]: I was thinking that a more granular implementation with `gr.Chatbot` would add more value to the guide. But there are no strong opinions here.

**Comment by yvrjsharma** [comment]
I've addressed all comments. Will go ahead and merge as discussed on Slack @abidlabs.

**Pull request merged**